File: scripts/validate_ids.py
import json
import re
from pathlib import Path


WEB_DIR = Path(__file__).resolve().parents[1]
ROOT = WEB_DIR.parents[1]
REGISTRY_DIR = ROOT / "_registry"


def read_json(path, fallback=None):
    if not path.exists():
        return fallback
    return json.loads(path.read_text(encoding="utf-8"))


def rel(path):
    return path.relative_to(ROOT).as_posix()


def as_list(value):
    if isinstance(value, list):
        return value
    if value in (None, ""):
        return []
    return [value]


def clean_scalar(value):
    return str(value or "").strip().strip("\"'")


def parse_frontmatter(text):
    match = re.match(r"^---\r?\n([\s\S]*?)\r?\n---\r?\n?", text)
    if not match:
        return {}
    meta = {}
    current = None
    for line in match.group(1).splitlines():
        item = re.match(r"^\s*-\s+(.*)$", line)
        if item and current:
            if not isinstance(meta.get(current), list):
                meta[current] = []
            meta[current].append(clean_scalar(item.group(1)))
            continue
        kv = re.match(r"^([A-Za-z0-9_]+):\s*(.*)$", line)
        if not kv:
            continue
        current = kv.group(1)
        meta[current] = [] if kv.group(2) == "" else clean_scalar(kv.group(2))
    return meta


def add_error(errors, path, message):
    errors.append(f"{rel(path)}: {message}")


def load_registries():
    company_registry = read_json(REGISTRY_DIR / "company_registry.json", {"companies": {}})
    product_registry = read_json(REGISTRY_DIR / "product_registry.json", {"products": {}})
    studio_registry = read_json(REGISTRY_DIR / "studio_registry.json", {"studios": {}})
    return {
        "companies": company_registry.get("companies", {}),
        "products": product_registry.get("products", {}),
        "studios": studio_registry.get("studios", {}),
    }


def validate_registry_internals(registries, errors):
    company_ids = set(registries["companies"].keys())
    product_ids = set(registries["products"].keys())
    studio_ids = set(registries["studios"].keys())
    product_registry_file = REGISTRY_DIR / "product_registry.json"
    studio_registry_file = REGISTRY_DIR / "studio_registry.json"

    for product_id, product in registries["products"].items():
        if product.get("product_id") != product_id:
            add_error(errors, product_registry_file, f"{product_id}: product_id field is {product.get('product_id')!r}")
        for company_id in as_list(product.get("developer_company_ids")) + as_list(product.get("publisher_company_ids")):
            if company_id not in company_ids:
                add_error(errors, product_registry_file, f"{product_id}: unknown company_id {company_id}")
        for studio_id in as_list(product.get("studio_ids")):
            if studio_id not in studio_ids:
                add_error(errors, product_registry_file, f"{product_id}: unknown studio_id {studio_id}")

    for studio_id, studio in registries["studios"].items():
        if studio.get("studio_id") and studio.get("studio_id") != studio_id:
            add_error(errors, studio_registry_file, f"{studio_id}: studio_id field is {studio.get('studio_id')!r}")
        for company_id in as_list(studio.get("parent_company_ids")):
            if company_id not in company_ids:
                add_error(errors, studio_registry_file, f"{studio_id}: unknown parent_company_id {company_id}")

    return company_ids, product_ids


def validate_financial_data(company_ids, errors):
    base = ROOT / "_financial_data"
    for path in base.rglob("*.json"):
        if "_templates" in path.parts:
            continue
        data = read_json(path, {})
        company_id = data.get("company_id")
        if company_id and company_id not in company_ids:
            add_error(errors, path, f"unknown company_id {company_id}")


def validate_markdown_refs(company_ids, product_ids, errors):
    for dir_name in ["_briefings", "_events", "_research_library"]:
        base = ROOT / dir_name
        if not base.exists():
            continue
        for path in base.rglob("*.md"):
            meta = parse_frontmatter(path.read_text(encoding="utf-8"))
            for company_id in as_list(meta.get("company_ids")):
                if company_id not in company_ids:
                    add_error(errors, path, f"unknown company_id in company_ids: {company_id}")
            if meta.get("company_id") and meta["company_id"] not in company_ids:
                add_error(errors, path, f"unknown company_id: {meta['company_id']}")
            for product_id in as_list(meta.get("product_ids")):
                if product_id not in product_ids:
                    add_error(errors, path, f"unknown product_id in product_ids: {product_id}")
            if meta.get("product_id") and meta["product_id"] not in product_ids:
                add_error(errors, path, f"unknown product_id: {meta['product_id']}")


def main():
    errors = []
    registries = load_registries()
    company_ids, product_ids = validate_registry_internals(registries, errors)
    validate_financial_data(company_ids, errors)
    validate_markdown_refs(company_ids, product_ids, errors)

    if errors:
        print(f"ID validation failed with {len(errors)} error(s):")
        for error in errors:
            print(f"- {error}")
        raise SystemExit(1)

    print(f"ID validation passed ({len(company_ids)} companies, {len(product_ids)} products)")


if __name__ == "__main__":
    main()

File: scripts/validate_ids.js
const fs = require("fs");
const path = require("path");

const WEB_DIR = path.resolve(__dirname, "..");
const ROOT = path.resolve(WEB_DIR, "..", "..");
const REGISTRY_DIR = path.join(ROOT, "_registry");

function readText(file) {
  return fs.readFileSync(file, "utf8");
}

function readJson(file, fallback = null) {
  if (!fs.existsSync(file)) return fallback;
  return JSON.parse(readText(file));
}

function rel(file) {
  return path.relative(ROOT, file).replace(/\\/g, "/");
}

function listFiles(dir, predicate = () => true) {
  if (!fs.existsSync(dir)) return [];
  const out = [];
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) out.push(...listFiles(full, predicate));
    if (entry.isFile() && predicate(full)) out.push(full);
  }
  return out;
}

function cleanScalar(value) {
  return String(value || "").trim().replace(/^["']|["']$/g, "");
}

function parseFrontMatter(text) {
  const match = text.match(/^---\r?\n([\s\S]*?)\r?\n---\r?\n?/);
  if (!match) return {};
  const meta = {};
  let current = null;
  for (const line of match[1].split(/\r?\n/)) {
    const listItem = line.match(/^\s*-\s+(.*)$/);
    if (listItem && current) {
      if (!Array.isArray(meta[current])) meta[current] = [];
      meta[current].push(cleanScalar(listItem[1]));
      continue;
    }
    const kv = line.match(/^([A-Za-z0-9_]+):\s*(.*)$/);
    if (!kv) continue;
    current = kv[1];
    meta[current] = kv[2] === "" ? [] : cleanScalar(kv[2]);
  }
  return meta;
}

function asArray(value) {
  if (Array.isArray(value)) return value;
  if (value === null || value === undefined || value === "") return [];
  return [value];
}

function addError(errors, file, message) {
  errors.push(`${rel(file)}: ${message}`);
}

function loadRegistries() {
  const companyRegistry = readJson(path.join(REGISTRY_DIR, "company_registry.json"), { companies: {} });
  const productRegistry = readJson(path.join(REGISTRY_DIR, "product_registry.json"), { products: {} });
  const studioRegistry = readJson(path.join(REGISTRY_DIR, "studio_registry.json"), { studios: {} });
  return {
    companies: companyRegistry.companies || {},
    products: productRegistry.products || {},
    studios: studioRegistry.studios || {},
  };
}

function validateRegistryInternals(registries, errors) {
  const companyIds = new Set(Object.keys(registries.companies));
  const productIds = new Set(Object.keys(registries.products));
  const studioIds = new Set(Object.keys(registries.studios));
  const productRegistryFile = path.join(REGISTRY_DIR, "product_registry.json");
  const studioRegistryFile = path.join(REGISTRY_DIR, "studio_registry.json");

  for (const [productId, product] of Object.entries(registries.products)) {
    if (product.product_id !== productId) {
      addError(errors, productRegistryFile, `${productId}: product_id field is ${JSON.stringify(product.product_id)}`);
    }
    for (const companyId of [...asArray(product.developer_company_ids), ...asArray(product.publisher_company_ids)]) {
      if (!companyIds.has(companyId)) addError(errors, productRegistryFile, `${productId}: unknown company_id ${companyId}`);
    }
    for (const studioId of asArray(product.studio_ids)) {
      if (!studioIds.has(studioId)) addError(errors, productRegistryFile, `${productId}: unknown studio_id ${studioId}`);
    }
  }

  for (const [studioId, studio] of Object.entries(registries.studios)) {
    if (studio.studio_id && studio.studio_id !== studioId) {
      addError(errors, studioRegistryFile, `${studioId}: studio_id field is ${JSON.stringify(studio.studio_id)}`);
    }
    for (const companyId of asArray(studio.parent_company_ids)) {
      if (!companyIds.has(companyId)) addError(errors, studioRegistryFile, `${studioId}: unknown parent_company_id ${companyId}`);
    }
  }

  return { companyIds, productIds, studioIds };
}

function validateFinancialData(companyIds, errors) {
  const base = path.join(ROOT, "_financial_data");
  for (const file of listFiles(base, (full) => /\.json$/i.test(full) && !full.includes(`${path.sep}_templates${path.sep}`))) {
    const data = readJson(file, {});
    if (data.company_id && !companyIds.has(data.company_id)) {
      addError(errors, file, `unknown company_id ${data.company_id}`);
    }
  }
}

function validateMarkdownRefs(companyIds, productIds, errors) {
  const targets = ["_briefings", "_events", "_research_library"];
  for (const dirName of targets) {
    const dir = path.join(ROOT, dirName);
    for (const file of listFiles(dir, (full) => /\.md$/i.test(full))) {
      const meta = parseFrontMatter(readText(file));
      for (const companyId of asArray(meta.company_ids)) {
        if (!companyIds.has(companyId)) addError(errors, file, `unknown company_id in company_ids: ${companyId}`);
      }
      if (meta.company_id && !companyIds.has(meta.company_id)) {
        addError(errors, file, `unknown company_id: ${meta.company_id}`);
      }
      for (const productId of asArray(meta.product_ids)) {
        if (!productIds.has(productId)) addError(errors, file, `unknown product_id in product_ids: ${productId}`);
      }
      if (meta.product_id && !productIds.has(meta.product_id)) {
        addError(errors, file, `unknown product_id: ${meta.product_id}`);
      }
    }
  }
}

function main() {
  const errors = [];
  const registries = loadRegistries();
  const { companyIds, productIds } = validateRegistryInternals(registries, errors);
  validateFinancialData(companyIds, errors);
  validateMarkdownRefs(companyIds, productIds, errors);

  if (errors.length) {
    console.error(`ID validation failed with ${errors.length} error(s):`);
    for (const error of errors) console.error(`- ${error}`);
    process.exitCode = 1;
    return;
  }

  console.log(`ID validation passed (${companyIds.size} companies, ${productIds.size} products)`);
}

main();
